akka-stream-testkit/src/main/scala/akka/stream/testkit/StreamTestKit.scala

```diff
@@ -0,0 +1,293 @@
+   */
+  class Probe[I] private[TestPublisher] ()(implicit system: ActorSystem) extends Publisher[I] {
+    private val probe: TestProbe = TestProbe()
+
+    /**
+     * Subscriber a given [[org.reactivestreams.Subscriber]] to this probe publisher.
+     */
+    def subscribe(subscriber: Subscriber[_ >: I]): Unit = {
+      val subscription: PublisherProbeSubscription[I] = new PublisherProbeSubscription[I](subscriber, probe)
+      probe.ref ! Subscribe(subscription)
+      subscriber.onSubscribe(subscription)
+    }
+
+    /**
+     * Expect a subscription.
+     */
+    def expectSubscription(): PublisherProbeSubscription[I] =
+      probe.expectMsgType[Subscribe].subscription.asInstanceOf[PublisherProbeSubscription[I]]
+
+    /**
+     * Expect demand from a given subscription.
+     */
+    def expectRequest(subscription: Subscription, n: Int): Unit = probe.expectMsg(RequestMore(subscription, n))
+
+    /**
+     * Expect no messages.
+     */
+    def expectNoMsg(): Unit = probe.expectNoMsg()
+
+    /**
+     * Expect no messages for a given duration.
+     */
+    def expectNoMsg(max: FiniteDuration): Unit = probe.expectNoMsg(max)
+
+    /**
+     * Receive messages for a given duration or until one does not match a given partial function.
+     */
+    def receiveWhile[T](max: Duration = Duration.Undefined, idle: Duration = Duration.Inf, messages: Int = Int.MaxValue)(f: PartialFunction[PublisherEvent, T]): immutable.Seq[T] =
+      probe.receiveWhile(max, idle, messages)(f.asInstanceOf[PartialFunction[AnyRef, T]])
+
+    def getPublisher: Publisher[I] = this
+  }
+
+  /**
+   * Demand tracking for [[TestPublisher.Probe]].
+   */
+  class Auto[T] private[TestPublisher] (initialPendingRequests: Long)(implicit system: ActorSystem) {
+
+    private var pendingRequests = initialPendingRequests
+    private lazy val subscription = probe.expectSubscription()
+
+    /**
+     * Underlying [[TestPublisher.Probe]].
+     */
+    val probe = TestPublisher.probe[T]()
+
+    /**
+     * Current pending requests.
+     */
+    def pending: Long = pendingRequests
+
+    def sendNext(elem: T): Unit = {
+      if (pendingRequests == 0) pendingRequests = subscription.expectRequest()
+      pendingRequests -= 1
+      subscription.sendNext(elem)
+    }
+
+    def sendNextDisregardDemand(elem: T): Unit = subscription.sendNext(elem)
+
+    def sendComplete(): Unit = subscription.sendComplete()
+
+    def sendError(cause: Exception): Unit = subscription.sendError(cause)
+
+    def expectRequest(): Long = subscription.expectRequest()
+
+    def expectCancellation(): Unit = subscription.expectCancellation()
+  }
+
+  sealed trait PublisherEvent extends DeadLetterSuppression
+  final case class Subscribe(subscription: Subscription) extends PublisherEvent
+  final case class CancelSubscription(subscription: Subscription) extends PublisherEvent
+  final case class RequestMore(subscription: Subscription, elements: Long) extends PublisherEvent
```

Comment:
If users see this perhaps `Request` would be better? `RequestMore` is old naming we had in some of the initial reactive streams proposals